Add NEXT100 Activity/Efficiency database

gondiaz opened the pull request:
This PR adds a new database to IC in similar terms as is done for detector's databases.

The new database is called NEXT100ACTDB, and is already placed in a dummy version at the mysql server ([mysql server](http://next.ific.uv.es/phpmyadmin/)). 

The database contains two tables, Activity and Efficiency with same column names. This database will be likely updated in the near future, and its purpose is to serve as an input to the MC Event-Mixer which will be also developed in subsequent PRs.

gondiaz committed:
Add new NEXT100DB tables in download

Cosmetics

gonzaponte commented on invisible_cities/database/load_db.py:
```diff
@@ -142,3 +142,17 @@ def PMTLowFrequencyNoise(db_file, run_number=1e5):
+@lru_cache(maxsize=10)
+def DataNext100Act():
+
+    filename = os.environ['ICTDIR'] + '/invisible_cities/database/localdb.NEXT100ACTDB.sqlite3'
```
the db file should be a parameter even if this is only for N100

jmbenlloch reviewed (approved):
This PR adds a new database with the activity assumptions.

gondiaz committed:
Add activity db reader

Rewrite activity reader to be more flexible

Alignment

gondiaz committed:
Add tests for activity reader

Rewrite database test

Fix typo in test

Rename df in test

Co-authored-by: Gonzalo <gonzaponte@gmail.com>

Add test for Efficiency table

Add test for version number request

Parametrize version in test

Avoid code repetition in test

Co-authored-by: Gonzalo <gonzaponte@gmail.com>

Improve redability

gondiaz committed:
Add new NEXT100DB tables in database

Update NEXT100DB

Update NEXT100DB

Update NEXT100DB

gonzaponte commented on invisible_cities/database/load_db.py:
```diff
@@ -142,3 +142,24 @@ def PMTLowFrequencyNoise(db_file, run_number=1e5):
+
+
+@lru_cache(maxsize=10)
+def MixerDataNext100(table, version=None):
```
I wouldn't name this function like this. Which program is going to use it shouldn't matter. I don't know what the best name is, maybe something like `RadioactivityData`.

gonzaponte commented on invisible_cities/database/load_db.py:
```diff
@@ -142,3 +142,24 @@ def PMTLowFrequencyNoise(db_file, run_number=1e5):
+def MixerDataNext100(table, version=None):
+
+    if table not in ["Activity", "Efficiency"]:
+        raise Exception("Unknow table type")
```
This looks to me like `table` shouldn't be a parameter, but rather something hardcoded. Specifically, you want to read both tables and return them both (or some combination of them).

gonzaponte commented on invisible_cities/database/load_db.py:
```diff
@@ -142,3 +142,24 @@ def PMTLowFrequencyNoise(db_file, run_number=1e5):
+
+    conn = sqlite3.connect(get_db("next100"))
+
+    if version is None:
+        version = pd.read_sql_query("SELECT MAX(Version) FROM Activity", conn).loc[0, "MAX(Version)"]
```
This looks like there is a more idiomatic way to do it, but I'm no expert. @jmbenlloch ?

gonzaponte replied:
Try to get an opinion from @jmbenlloch. It's not critical, but if there is a better way we should use it.

jmbenlloch replied:
Looks good to me. The SQL part has to be that way, the pandas part is just to select the column and the value.

gonzaponte commented on invisible_cities/database/load_db.py:
```diff
@@ -142,3 +142,24 @@ def PMTLowFrequencyNoise(db_file, run_number=1e5):
+            '''.format(version)
+    activity = pd.read_sql_query(query, conn)
+    conn.close()
+    return activity.drop(columns="MAX(Version)")
```
I've seen that the efficiency table is currently empty and that's probably why you only work with activity. Can we fill it with meaningless values so the code can already be prepared for it?

gonzaponte commented on invisible_cities/database/load_db_test.py:
```diff
@@ -178,3 +179,24 @@ def test_pmt_noise_frequencies(db):
+            '''
+
+    for (G4Volume, Isotope), act in df.groupby(["G4Volume", "Isotope"]):
+        act_ = pd.read_sql_query(query.format(G4Volume, Isotope), conn)
```
being picky here. In python, we use capitalized names for classes. we kinda made an exception with the `DB` import and the database in general, but we shouldn't get used to it. Rename to `g4volume, isotope` or similar.

gonzaponte replied:
Also, rename `act` and  `act_` to something more explicit like `requested` and `full`.

gonzaponte commented on invisible_cities/database/load_db_test.py:
```diff
@@ -178,3 +179,24 @@ def test_pmt_noise_frequencies(db):
+
+    for (G4Volume, Isotope), act in df.groupby(["G4Volume", "Isotope"]):
+        act_ = pd.read_sql_query(query.format(G4Volume, Isotope), conn)
+        act_ = act_[(act_["Version"] == act_["Version"].max())]
```
```suggestion
        act_ = act_[act_.Version == act_.Version.max()]
```

gonzaponte commented on invisible_cities/database/load_db_test.py:
```diff
@@ -178,3 +179,24 @@ def test_pmt_noise_frequencies(db):
+        act_ = pd.read_sql_query(query.format(G4Volume, Isotope), conn)
+        act_ = act_[(act_["Version"] == act_["Version"].max())]
+
+        np.testing.assert_allclose(act["TotalActivity"], act_["TotalActivity"])
```
```suggestion
        np.testing.assert_allclose(act.TotalActivity, act_.TotalActivity)
```
Getting columns as attributes is more readable.

gonzaponte commented on invisible_cities/database/download.py:
```diff
@@ -77,12 +72,18 @@ def loadDB(dbname : str):
-
-
+        for dbname in table_dict:
+            loadDB(dbname, table_dict[dbname])
```
```suggestion
        for dbname, tables in table_dict.items():
            loadDB(dbname, tables)
```

gonzaponte commented on invisible_cities/database/load_db.py:
```diff
@@ -142,3 +142,24 @@ def PMTLowFrequencyNoise(db_file, run_number=1e5):
+    if table not in ["Activity", "Efficiency"]:
+        raise Exception("Unknow table type")
+
+    conn = sqlite3.connect(get_db("next100"))
```
This needs to be callable from any database. Right now we have only NEXT100, but it should be extensible to any other.

gonzaponte reviewed (commented):
First round. Please run `pyflakes` after you address these comments to check for unused variables.

gonzaponte commented on invisible_cities/database/load_db_test.py:
```diff
@@ -178,3 +179,26 @@ def test_pmt_noise_frequencies(db):
+        req = pd.read_sql_query(query.format(g4volume, isotope), conn)
+        req = req[req.Version == req.Version.max()]
+
+        np.testing.assert_allclose(req.TotalActivity, full.TotalActivity)
```
The naming is the other way around, right? `df` contains the data you requested (the one you are testing). `pd.read_sql_query(query.format(g4volume, isotope), conn)` gives you the full output, not filtered by version. Perhaps `full` wasn't the best name suggestion, how about
```suggestion
    for (g4volume, isotope), req in df.groupby(["G4Volume", "Isotope"]):
        all_versions = pd.read_sql_query(query.format(g4volume, isotope), conn)
        max_version = all_versions[all_versions.Version == all_versions.Version.max()]

        np.testing.assert_allclose(req.TotalActivity, max_version.TotalActivity)
```

Also, why the groupby? Is it because the data may come in different orders? If so, you may want to sort both dataframes and avoid the loop using `assert_dataframes_equal`.

Finally, this test checks that without version number you get the highest one, and that's great, but you should write another one forcing a specific version number and check that you get the correct result.

gondiaz replied:
Not sure what you mean with different data orders. The objective is to avoid loading the whole database table, first selecting the data for a single G4Volume and Isotope and then explicitly keep the TotalActivity of the highest version number. This tests whether the sql-query for the whole database table in the `RadioactivityData` function is doing what we want.

I assumed that the test for a version number is uneccessary since the query in `RadioactivityData` is the same and we may only have a `version=0` (which may be unchanged for some time) once a non-dummy database is loaded. I can add the test for `version=0` if you found it neccessary now, or wait for changed in the database in the future.

gonzaponte replied:
> The objective is to avoid loading the whole database table, 

Ok. Loading the full table is probably a non-issue, but I agree that there is no need to do that. I was thinking that requesting so many queries to mysql could make it run slow. If that's not the case I'm ok with it.

>  I can add the test for version=0 if you found it neccessary now, or wait for changed in the database in the future.

It doesn't take long and we will probably forget about adding that test, so better do it now that we are focused on it.

We need to make it work for many versions so we I would add also version 2 updating other values than in version 1 to the activity table and eventually do the same with the efficiency table.

Once we do this with a dummy database, we can sleep in peace knowing that the code is ready to handle the real database.

gondiaz replied:
I have added the test for Efficiency table and for the version number. The new test does it for `version = 0`, otherwise, for `version > 0` may give future problem since the first non-dummy database version could be left unmodified during some time at `version = 0` until some update is needed.

I am not aware if the sql querys are slower (my unfounded opinion would say that no, since sqlite should be designed to allow fast data selection), but the tests are fast anyway.

gonzaponte replied:
My point was actually that we should add a test (or many) requesting versions 0, 1, and 2 _now_ because we have this dummy database with multiple versions. Once we load the real database we won't have multiple versions so the tests won't tell us much. Nonetheless, we will be confident that it works because we tested it before with this dummy database.

In principle, the test shouldn't fail for versions 1 and 2 even if they don't exist in the database because it should take the last available version. Therefore, I think you can simply parametrize it.

gonzaponte commented on invisible_cities/database/load_db_test.py:
```diff
@@ -178,3 +179,64 @@ def test_pmt_noise_frequencies(db):
+    for (g4volume, isotope), req in efficiency.groupby(["G4Volume", "Isotope"]):
+        all_versions = pd.read_sql_query(query.format(g4volume, isotope), conn)
+        max_version = all_versions[all_versions.Version == all_versions.Version.max()]
+        np.testing.assert_allclose(req.MCEfficiency, max_version.MCEfficiency)
```
loop to avoid code repetition? Same for the other test. Review alignment and I think we are done.

gondiaz replied:
Not sure how is the best way to do it or if it is clearer, since the column name is different for Activity table (TotalActivity) and Efficiency (MCEfficiency). Also, (g4volume, isotope) might be missing, we are not testing that both tables contain all the combinations, which for example is not the case in the current dummy tables.

gonzaponte replied:
I was referring to something like
```suggestion
    for table, table_name, column in zip((      activity ,    efficiency ),
                                         (     "Activity",   "Efficiency"),
                                         ("TotalActivity", "MCEfficiency")):

        query = "SELECT * FROM {0} WHERE G4Volume = '{1}' AND Isotope = '{2}'"
        for (g4volume, isotope), req in table.groupby(["G4Volume", "Isotope"]):
            all_versions = pd.read_sql_query(query.format(table_name, g4volume, isotope), conn)
            max_version = all_versions[all_versions.Version == all_versions.Version.max()]
            np.testing.assert_allclose(req[column], max_version[column])
```
I think the lack of repetition compensates for the slightly less explicit code.

I don't know what do you mean by
> (g4volume, isotope) might be missing

gondiaz replied:
Ok, that looks good to me. I was discarting to do the column call like `req[column]` since you requested above the change of sintax `df["column_name"]` to `df.column_name`.

gonzaponte replied:
It's a bit of a trade-off. I definitely prefer using attributes, but that's not so handy when the name is stored in a variable. So as a general rule, use attribute access for hard-coded values and item access for variables.

gonzaponte commented on invisible_cities/database/load_db.py:
```diff
@@ -142,3 +142,24 @@ def PMTLowFrequencyNoise(db_file, run_number=1e5):
+               GROUP BY G4Volume, Isotope
+            '''
+    activity   = pd.read_sql_query(query.format("TotalActivity", "Activity", version), conn)
+    efficiency = pd.read_sql_query(query.format( "MCEfficiency","Efficiency", version), conn)
```
```suggestion
    activity   = pd.read_sql_query(query.format("TotalActivity",   "Activity", version), conn)
    efficiency = pd.read_sql_query(query.format( "MCEfficiency", "Efficiency", version), conn)
```

gonzaponte commented on invisible_cities/database/load_db_test.py:
```diff
@@ -178,3 +179,47 @@ def test_pmt_noise_frequencies(db):
+                                         (     "Activity",   "Efficiency"),
+                                         ("TotalActivity", "MCEfficiency")):
+        for (g4volume, isotope), req in table.groupby(["G4Volume", "Isotope"]):
+            all_versions = pd.read_sql_query(query.format(table_name, g4volume, isotope), conn)
```
I think defining the query string within the loops will be more readable since you can use the f-string formatting.

gonzaponte reviewed (approved):
This PR adds new tables to the database and implements the functionality to load them within IC. It also provides a couple of tests to demonstrate and check the functionality.

Great job.
